ntp-proto/src/packet.rs

```diff
@@ -113,25 +114,162 @@ pub struct NtpPacket<'a> {
+    },
+    NtsEncryptedField {
+        nonce: Cow<'a, [u8]>,
+        plaintext: Cow<'a, [u8]>,
+    },
+
+    Unknown {
+        typeid: u16,
+        data: Cow<'a, [u8]>,
+    },
+}
+
+impl<'a> std::fmt::Debug for ExtensionField<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::UniqueIdentifier(arg0) => f.debug_tuple("UniqueIdentifier").field(arg0).finish(),
+            Self::NtsCookie(arg0) => f.debug_tuple("NtsCookie").field(arg0).finish(),
+            Self::NtsCookiePlaceholder { body_length } => f
+                .debug_struct("NtsCookiePlaceholder")
+                .field("body_length", body_length)
+                .finish(),
+            Self::NtsEncryptedField { nonce, plaintext } => f
+                .debug_struct("NtsEncryptedField")
+                .field("nonce", nonce)
+                .field("plaintext", plaintext)
+                .finish(),
+            Self::Unknown { typeid, data } => f
+                .debug_struct("Unknown")
+                .field("typeid", typeid)
+                .field("length", &data.len())
+                .field("data", data)
+                .finish(),
+        }
+    }
+}
+
+pub const fn next_multiple_of(lhs: usize, rhs: usize) -> usize {
+    match lhs % rhs {
+        0 => lhs,
+        r => lhs + (rhs - r),
+    }
 }
 
 impl<'a> ExtensionField<'a> {
     const MINIMUM_SIZE: usize = 16;
 
     fn into_owned(self) -> ExtensionField<'static> {
+        use ExtensionField::*;
+
         match self {
-            ExtensionField::Unknown { typeid, data } => ExtensionField::Unknown {
+            Unknown { typeid, data } => Unknown {
                 typeid,
                 data: Cow::Owned(data.into_owned()),
             },
+            UniqueIdentifier(data) => UniqueIdentifier(Cow::Owned(data.into_owned())),
+            NtsCookie(data) => NtsCookie(Cow::Owned(data.into_owned())),
+            NtsCookiePlaceholder { body_length } => NtsCookiePlaceholder { body_length },
+            NtsEncryptedField { nonce, plaintext } => NtsEncryptedField {
+                nonce: Cow::Owned(nonce.into_owned()),
+                plaintext: Cow::Owned(plaintext.into_owned()),
+            },
         }
     }
 
-    fn serialize<W: std::io::Write>(&self, w: &mut W) -> std::io::Result<()> {
+    pub fn key_exchange_signature(nonce: &'a Nonce) -> Self {
+        // the real plaintext is the packet up to the start of this extension field
+        // it is inserted implicitly by the encoder
+        ExtensionField::NtsEncryptedField {
+            nonce: nonce.as_slice().into(),
+            plaintext: [].as_slice().into(),
+        }
+    }
+
+    pub fn serialize_without_encryption(&self, w: &mut Cursor<&mut [u8]>) -> std::io::Result<()> {
+        use aes_siv::{aead::KeyInit, Key};
+
+        let cipher = Aes256SivAead::new(Key::<Aes256SivAead>::from_slice([0; 64].as_slice()));
+
+        self.serialize(w, &cipher)
+    }
```

Comment:
This feels like a very insecure function to have available as part of the public non-testing interface. Why is it needed?